File: Problems/238. Product of Array Except Self.cpp
#include<vector>
#include<iostream>
using namespace std;


class Solution {
    public:
        vector<int> productExceptSelf(vector<int>& nums) {
            
            int total = 1;
            int totalWithoutZero = 1;
            int zeroCounter = 0;

            vector<int> result;

            for(int i=0;i<nums.size();i++){
                total = total*nums[i];
                if(nums[i]==0){
                    zeroCounter++;
                }
                if(nums[i] != 0){
                    totalWithoutZero = totalWithoutZero*nums[i];
                }
            }            


            for(int i=0;i<nums.size();i++){

                if(nums[i] != 0){
                    result.push_back(total/nums[i]);
                }
                else if (zeroCounter > 1 && nums[i] == 0){
                    result.push_back(0);
                    // result.push_back(total/nums[i]);
                }
                else{
                    result.push_back(totalWithoutZero);

                }
            }


            return result;  
         }
};

void printVector(vector<int>& nums){
    cout << "-> ";
    for(int i=0;i<nums.size();i++){
        cout << " " << nums[i] << " ";
    }
    cout << endl;
}

int main(){
    Solution solution;

    vector<int> input = {0,0};

    vector<int> answer = solution.productExceptSelf(input);
    printVector(answer);

}

File: 238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        prefix = [0]*len(nums)
        postfix = [0]*len(nums)
        result = [0]*len(nums)
        cur = 1

        #Calculate prefix 
        for index in range(len(nums)):
            cur = cur*nums[index]
            prefix[index] = cur

        cur = 1
        #Calculate postfix
        for index in range(len(nums)-1,-1,-1):
            cur = cur*nums[index]
            postfix[index] = cur

        for index in range(len(nums)):
            if index != 0 and index != len(nums)-1:
                result[index] = prefix[index-1]*postfix[index+1]
            elif index == 0:
                result[index] = postfix[index+1]
            elif index == len(nums)-1:
                result[index] = prefix[index-1]
        
        return result